### PR TITLE
Allow URLs with custom protocols in links 

### LINKED ISSUE #14304: Allow URLs with custom protocols to be linkable using Link plugin
## 📝 Provide a description of the improvement
The improvement I'd like to propose is adding the ability to customize which URL protocols the Link plugin considers safe so that URLs using them are properly linked in the Link Balloon and take the user to the intended destination upon clicking. 

_How the feature works now and what you'd like to change_?

Currently, the Link Balloon relies a `SAFE_URL` regex to determine if a link, upon click, should take the user to the set URL, or to `#` 

```js
const SAFE_URL = /^(?:(?:https?|ftps?|mailto):|[^a-z]|[a-z+.-]+(?:[^a-z+.:-]|$))/i;
```

Because the list of "safe" protocols are baked into the `SAFE_URL` regex, URLs using uncommon protocols such as `tel`, cannot be linked.

In order to address this, there could be an option added to the link configuration that might look something like this: 

```js
<CKEditor
  config={{
    link: {
      safeUrlProtocols: 'https?|ftps?|mailto|tel' // or alternatively ['http', 'ftp', 'mailto', 'tel']
    }
  }}
/>
```

With this approach, the regex currently used to determine if a URL is safe might be replaced with a function that receives `safeUrlProtocols` to dynamically build and return a regex. 

## 📃 Other details

* Browser: All
* OS: All
* CKEditor version: All
* Installed CKEditor plugins: Link, AutoLink

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by map-r**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature (link): Added the ability to specify allowed URL protocols by introducing the `link.allowedProtocols` configuration property. Closes #14304.

---

### Additional information

I'm unsure which is better: whether the `allowedProtocols` should override the default protocols list or merge with it, when set.

Kept the old regex intact (only added replace tag for the custom list), but there's definitely room for improvement there, I described it at https://github.com/ckeditor/ckeditor5/issues/15915

### CODE REVIEW

**Comment by map-r** [comment]
One missed and just spotted: `file:` protocol, I think it should be also the default.

**Comment by Witoso** [comment]
> One missed and just spotted: `file:` protocol, I think it should be also the default.

It's important for sure (#8248), but considering that our editor is mostly used as a web editor in (shared) apps, I doubt linking to specific local files should be the default.

File: packages/ckeditor5-link/src/linkconfig.ts [hunk 1]
```diff
@@ -9,6 +9,44 @@
 
 import type { ArrayOrItem } from 'ckeditor5/src/utils.js';
 
+/**
+ * This is a list of most commonly used protocols to make the link custom configuration a bit easier.
+ * The {@link module:link/linkconfig~LinkConfig#allowedProtocols `allowedProtocols` } list
+ * can also contain other values than enlisted here.
+ */
+export enum COMMON_PROTOCOLS {
```
Comment (DawidKossowski) [inline packages/ckeditor5-link/src/linkconfig.ts:17]: I am not sure if we need this enum. Who will use it and why? 
I'm pretty sure that if someone wants to use custom protocols, they exactly know which one.

File: packages/ckeditor5-link/src/linkconfig.ts [hunk 2]
```diff
@@ -48,6 +86,28 @@ export interface LinkConfig {
...
+	 * You can use the {@link module:link/linkconfig~COMMON_PROTOCOLS common protocols list} for help.
+	 *
+	 * **Note:** Use this with caution and at your own risk - adding unsafe protocols like `javascript:`
+	 * can result in serious security vulnerabilities!
+	 *
+	 * ```ts
+	 * ClassicEditor
+	 * 	.create( editorElement, {
+	 * 		link: {
+	 * 			allowedProtocols: ['http', 'https', 'ftp', 'tel', 'mailto', 'ssh']
```
Comment (DawidKossowski) [inline packages/ckeditor5-link/src/linkconfig.ts:101]: Please use our code style (whitespaces).

**Review by DawidKossowski — CHANGES_REQUESTED** [review]
Two things are still missing:
1. There is no information in the docs.
2. The tests you added barely cover any error cases.
3. You didn't add any tests for `linkediting` and `linkactionsview`, but you expanded the constructor, and so on.

File: packages/ckeditor5-link/src/linkconfig.ts (hunk 1 above)
Comment (scofalik) [reply packages/ckeditor5-link/src/linkconfig.ts:17]: I think this is not needed and people will not use it. They will go with `allowedProtocols: [ 'http', 'ftp', 'foobar' ]`.

File: packages/ckeditor5-link/src/utils.ts [hunk 3]
```diff
@@ -19,16 +19,27 @@ import type {
...
+	DEFAULT_LINK_PROTOCOLS,
+	type LinkDecoratorAutomaticDefinition,
+	type LinkDecoratorDefinition,
+	type LinkDecoratorManualDefinition
 } from './linkconfig.js';
 
 import { upperFirst } from 'lodash-es';
 
 const ATTRIBUTE_WHITESPACES = /[\u0000-\u0020\u00A0\u1680\u180E\u2000-\u2029\u205f\u3000]/g; // eslint-disable-line no-control-regex
-const SAFE_URL = /^(?:(?:https?|ftps?|mailto):|[^a-z]|[a-z+.-]+(?:[^a-z+.:-]|$))/i;
```
[the remark this answers is not in the document]
Comment (map-r) [reply packages/ckeditor5-link/src/utils.ts:31]: https://www.notion.so/9fc0ee1f68be4a43bb535d30574f9e2c?v=2c74ccd33c6e4446bd858f5368f9932c&p=3382338c77594e06953a3d4879bd2400&pm=s
check about the old regex, the problem is the second and next cases which probably intended to allow relative urls but as Piotr admitted in a private chat, there probably should be an update to this old regex.
also many parts are drawn from the autolink regex, which is also different.

File: packages/ckeditor5-link/src/linkconfig.ts (hunk 1 above)
Comment (map-r) [reply packages/ckeditor5-link/src/linkconfig.ts:17]: okay, a proposition as said ;) will have only literals

File: packages/ckeditor5-link/src/linkconfig.ts [hunk 4]
```diff
@@ -9,6 +9,44 @@
...
+	VIBER = 'viber',
+	JABBER = 'jabber',
+	ZOOMMTG = 'zoommtg',
+	ZOOMUS = 'zoomus'
+}
+
+/**
+ * @internal
+ */
+export const DEFAULT_LINK_PROTOCOLS = [
```
[the remark this answers is not in the document]
Comment (map-r) [reply packages/ckeditor5-link/src/linkconfig.ts:41]: First of all, I'm not sure what type of breaking changes when you extend the list of protocols, not omitting any of the old ones.
Second, well, I did this because this is how I interpreted the connected tickets, for instance:
https://github.com/ckeditor/ckeditor5/issues/11855

but probably they will simply need to make custom configs then. fine both ways.

[the remark this answers is not in the document]
Comment (map-r) [reply packages/ckeditor5-link/src/linkconfig.ts:41]: added were sftp, tel and sms, as I took it from the pinned related issues.

[the remark this answers is not in the document]
Comment (DawidKossowski) [reply packages/ckeditor5-link/src/linkconfig.ts:41]: Just to clarify, changing default behaviors, whether we're limiting or expanding functionality, can lead to breaking changes. For example, if someone doesn't want to support a specific protocol (which was not previously supported by default), they'll have to modify their code. It might not be a big deal, but it still provides a minor breaking change. Remember that different clients/users might have different situations or requirements in their applications.

Secondly, as Szymon mentioned, that was not the point of this task.

[the remark this answers is not in the document]
Comment (map-r) [reply packages/ckeditor5-link/src/linkconfig.ts:41]: Okay then, just to be completely clear. I thought it **is** a scope of this task just because of how the linked issues about several protocols (which were clearly request to add support for them) were described and concluded. But it's also true that in the main #14304 we are focused only on adding of the customization option. So will keep the original list countet to my prior intuition.

[the remark this answers is not in the document]
Comment (map-r) [reply packages/ckeditor5-link/src/linkconfig.ts:41]: @Witoso Could you take a look? I think idea of adding or not adding any of the requested link protocols to the default set deserves one more opinion, if you will.

File: packages/ckeditor5-link/src/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (map-r) [reply packages/ckeditor5-link/src/utils.ts:31]: If this is not the drill, then please let me know what process should it follow to meet safety requirements (which I have doubts that they've been met in the old way), and we can shelve this off to another issue, keeping old regex contents intact.

**Comment by map-r** [comment]
What's done post review:
- removed enum of common protocols
- added tests requested

@DawidKossowski To reply to general points raised:
1. to be precise, I've added the description on the `allowedProtocols` for API docs, so this is not quite true, what you mean are the guides/feature docs, from that I'm getting definite feeling that all of the things config-related require a section in docs, will keep that in mind onwards,
2. This is not really helpful feedback, in terms, I'm not sure what would you expect to be added, besides a junk invalid URL examples (one added), probably the fact that I've touched upon this thing makes you demand writing even more erroring tests, however up to this point we were fine with the current set (in safe url checking at least). I can either: a) expand on it and look for the top problematic URLs, b) discard this and shelve this off to another time, another issue.
3. right, done, I'm not sure why I omitted it, perhaps I was looking into another set of changes and overlooked that. added

File: packages/ckeditor5-link/src/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (map-r) [reply packages/ckeditor5-link/src/utils.ts:31]: I believe two factors were major in my idea of changing it, being:
- unable to link `localhost`,
- able to link some foreign protocols which contain number or dot or else (there are some like this in fact)

I wanted to consult this on a PR level, but I guess I was wrong about it. I'll consult such stuff more early on then.

**Comment by DawidKossowski** [comment]
1. You wrote that this is not true, but it actually is. Just for clarification, when I say 'docs,' I'm referring to user guides rather than API documentation, which are available here: https://ckeditor.com/docs/ckeditor5/latest/features/link.html 
2. I meant the different incorrect cases combined with the new config option.

**Comment by map-r** [comment]
@DawidKossowski Docs subsection added, and I'll just once more point to the discussion about changing regex, that I don't wan't to potentially waste more time, so waiting on some decisions here about whether we conduct security tests (so I shall consult right people for this) or do we put this away now and perhaps open a new ticket to improve the regex?

File: packages/ckeditor5-link/tests/linkediting.js [hunk 5]
```diff
@@ -101,7 +101,7 @@ describe( 'LinkEditing', () => {
 			// So let's simulate the `keydown` event.
 			editor.editing.view.document.fire( 'keydown', {
 				keyCode: keyCodes.arrowright,
-				preventDefault: () => {},
+				preventDefault: () => { },
```
Comment (DawidKossowski) [inline packages/ckeditor5-link/tests/linkediting.js:104]: Please revert all unnecessary changes (whitespaces).

File: packages/ckeditor5-link/src/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (DawidKossowski) [reply packages/ckeditor5-link/src/utils.ts:31]: I would like to process this change as a separate task to narrow down the current scope. 

Therefore, please keep the old regexp and create the follow-up task.

**Review by DawidKossowski — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-link/src/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (map-r) [reply packages/ckeditor5-link/src/utils.ts:31]: here, reverted the thing but still the Regex has a protocols replacement tag in it so we can embed list of protocols as in the task.
